cmd/test/makefiles/speed.go

```diff
@@ -0,0 +1,245 @@
+package makefiles
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"time"
+
+	"github.com/rclone/rclone/cmd"
+	"github.com/rclone/rclone/cmd/test"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/cache"
+	"github.com/rclone/rclone/fs/config/flags"
+	"github.com/rclone/rclone/fs/operations"
+	"github.com/rclone/rclone/fs/sync"
+	"github.com/rclone/rclone/lib/atexit"
+	"github.com/rclone/rclone/lib/random"
+	"github.com/spf13/cobra"
+)
+
+var (
+	// Flags
+	testTime = fs.Duration(15 * time.Second)
+	fcap     = 100
+	small    = fs.SizeSuffix(1024)
+	medium   = fs.SizeSuffix(10 * 1024 * 1024)
+	large    = fs.SizeSuffix(1024 * 1024 * 1024)
+	useJSON  = false
+)
+
+func init() {
+	test.Command.AddCommand(speedCmd)
+
+	speedFlags := speedCmd.Flags()
+	flags.FVarP(speedFlags, &testTime, "test-time", "", "Length for each test to run", "")
+	flags.IntVarP(speedFlags, &fcap, "file-cap", "", fcap, "Maximum number of files to use in each test", "")
+	flags.FVarP(speedFlags, &small, "small", "", "Size of small files", "")
+	flags.FVarP(speedFlags, &medium, "medium", "", "Size of medium files", "")
+	flags.FVarP(speedFlags, &large, "large", "", "Size of large files", "")
+	flags.BoolVarP(speedFlags, &useJSON, "json", "", useJSON, "Output only results in JSON format", "")
+
+}
+
+var speedCmd = &cobra.Command{
+	Use:   "speed <remote> [flags]",
+	Short: `Run a speed test to the remote`,
+	Long: `Run a speed test to the remote.
+
+	This command runs a series of uploads and downloads to the remote, measuring
+	and printing the speed of each test using varying file sizes and numbers of
+	files.
+
+	Test time can be innaccurate with small file caps and large files. As it
+	uses the results of an initial test to determine how many files to use in
+	each subsequent test.
+
+	It is recommended to use -q flag for a simpler output.
+
+	**NB** This command will create and delete files on the remote in a randomly
+	named directory which should be tidied up after.
+
+	You can use the --json flag to only print the results in JSON format.`,
+	Annotations: map[string]string{
+		"versionIntroduced": "v1.72",
+	},
+	RunE: func(command *cobra.Command, args []string) error {
+		ctx := command.Context()
+		cmd.CheckArgs(1, 1, command, args)
+		commonInit()
+
+		// initial test
+		size := fs.SizeSuffix(1024 * 1024)
+		if !useJSON {
+			fmt.Printf("Running initial test for 4 files of size %v\n", size)
+		}
+		stats, err := speedTest(ctx, 4, size, args[0])
+		if err != nil {
+			return fmt.Errorf("speed test failed: %w", err)
+		}
+
+		var results []*Stats
+
+		// main tests
+		for _, size := range []fs.SizeSuffix{small, medium, large} {
+			combinedSpeed := stats.Upload.Speed + stats.Download.Speed
```

Comment:
I don't think this calculation is quite right... Maybe this should be the average speed if the intention is that the total test for upload and download should hit the `testTime`?

It might be best to do the upload and download calculations separately as they are likely to be quite different.